File: BluePay iOS SDK-ObjC/BluePay iOS SDK/BluePayRequest.m
//
//  BluePayRequest.m
//  BluePay Apple Pay SDK
//
//  Created by Justin Slingerland on 3/16/15.
//  Copyright (c) 2015 BluePay Processing, Inc. All rights reserved.
//

#import "BluePayRequest.h"

@implementation BluePayRequest

+ (void)Post:(NSData*)token bluepaySetup:(NSDictionary*) bpSetup customer:(NSDictionary*)customerInfo handler:(void(^)(NSString *))completion
{
    if (![bpSetup[@"TransType"]  isEqual: @"SALE"] && ![bpSetup[@"TransType"]  isEqual: @"AUTH"])
        completion(@"Transaction type must be either SALE or AUTH");
    
    NSError *error = nil;
    id object = [NSJSONSerialization
                 JSONObjectWithData:token
                 options:0
                 error:&error];
    if (error)
    {
        NSLog(@"There was an error in retrieving the Apple Pay token");
        completion(@"There was an error in retrieving the Apple Pay token");
    }
    
    NSDictionary *results = object;

    // Create POST string to send to BluePay
    NSMutableString *post = [NSMutableString string];
        [post appendString:@"ACCOUNT_ID=%@"];
        [post appendString:@"&MODE=%@"];
        [post appendString:@"&TRANS_TYPE=%@"];
        [post appendString:@"&NAME1=%@"];
        [post appendString:@"&NAME2=%@"];
        [post appendString:@"&ADDR1=%@"];
        [post appendString:@"&CITY=%@"];
        [post appendString:@"&STATE=%@"];
        [post appendString:@"&ZIP=%@"];
        [post appendString:@"&COUNTRY=%@"];
        [post appendString:@"&PHONE=%@"];
        [post appendString:@"&EMAIL=%@"];
    
    NSMutableString *postString = [NSMutableString stringWithFormat:post,bpSetup[@"AccountID"],bpSetup[@"TransMode"],bpSetup[@"TransType"],customerInfo[@"FirstName"],customerInfo[@"LastName"],customerInfo[@"Street"],customerInfo[@"City"],customerInfo[@"State"],customerInfo[@"ZIP"],customerInfo[@"Country"],customerInfo[@"Phone"],customerInfo[@"Email"]];
    [post setString:@""];
    if (results[@"header"]) {
        NSString *tps = [BluePay calcTPS:bpSetup[@"SecretKey"] accID:bpSetup[@"AccountID"] transType:bpSetup[@"TransType"] transAmount:@"" fullName:customerInfo[@"FirstName"] paymentAcct:@""];
        [post appendString:@"&TAMPER_PROOF_SEAL=%@"];
        [post appendString:@"&APPLE_EPK=%@"];
        [post appendString:@"&APPLE_DATA=%@"];
        [post appendString:@"&APPLE_SIG=%@"];
        [postString appendString:[NSString stringWithFormat:post,tps,[results[@"header"][@"ephemeralPublicKey"] stringByAddingPercentEncodingForRFC3986],[results[@"data"] stringByAddingPercentEncodingForRFC3986],[results[@"signature"] stringByAddingPercentEncodingForRFC3986]]];
    } else if (results[@"EncryptedTrack1"]) {
        NSString *tps = [BluePay calcTPS:bpSetup[@"SecretKey"] accID:bpSetup[@"AccountID"] transType:bpSetup[@"TransType"] transAmount:results[@"Amount"] fullName:results[@"FirstName"]  paymentAcct:@""];
        [post appendString:@"&TAMPER_PROOF_SEAL=%@"];
        [post appendString:@"&AMOUNT=%@"];
        [post appendString:@"&TRACK1_ENC=%@"];
        [post appendString:@"&TRACK1_EDL=%@"];
        [post appendString:@"&KSN=%@"];
        [postString appendString:[NSString stringWithFormat:post,tps,customerInfo[@"Amount"],results[@"EncryptedTrack1"],results[@"Track1Length"],results[@"KSN"]]];
    } else {
        NSString *tps = [BluePay calcTPS:bpSetup[@"SecretKey"] accID:bpSetup[@"AccountID"] transType:bpSetup[@"TransType"] transAmount:results[@"Amount"] fullName:results[@"FirstName"] paymentAcct:results[@"CardNumber"]];
        [post appendString:@"&TAMPER_PROOF_SEAL=%@"];
        [post appendString:@"&AMOUNT=%@"];
        [post appendString:@"&PAYMENT_ACCOUNT=%@"];
        [post appendString:@"&CARD_EXPIRE=%@"];
        [postString appendString:[NSString stringWithFormat:post,tps,customerInfo[@"Amount"],results[@"CardNumber"],results[@"CardExpirationDate"]]];
    }
    // POST to the bp20post API of the BluePay gateway
    NSString *postURL = @"https://secure.bluepay.com/interfaces/bp20post";
    NSMutableURLRequest *postRequest = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:postURL]];
    
    [postRequest setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    [postRequest setValue:@"BluePay iOS SDK" forHTTPHeaderField:@"User-Agent"];
    [postRequest setHTTPMethod:@"POST"];
    [postRequest setHTTPBody:[postString dataUsingEncoding:NSUTF8StringEncoding]];
    
    // Create the POST request and specify its body data
    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:config];

    //NSData *responseData = [NSURLConnection sendSynchronousRequest:postRequest returningResponse:&response error:&err];
    NSURLSessionDataTask *postToBP = [session dataTaskWithRequest:postRequest completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        //NSLog(@"Response:%@ %@\n", response, error);
        if(error == nil)
        {
            NSString * response = [[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding];
            //NSLog(@"Data = %@",response);
            completion(response);
            return;
        } else {
            completion([error localizedDescription]);
        }
    }];
        //completion(@"There was an error in retrieving the transaction details from the BluePay gateway.");
    [postToBP resume];
    return;
    //NSString *responseBody = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
    // Return the body of the HTTPS POST response from BluePay
    //return responseBody;
}

@end

File: BluePay iOS SDK-Swift/BluePay iOS SDK-Swift/BluePayRequest.swift
import Foundation
class BluePayRequest: NSObject {
    class func Post(bluepaySetup bpSetup: [String: String], customer customerInfo: [String: String], postCompleted : (succeeded: Bool, msg: String) -> ()) {
        if !bpSetup["TransType"]!.isEqual("SALE") && !bpSetup["TransType"]!.isEqual("AUTH") {
            postCompleted(succeeded: false, msg: "Error: Transaction type must be either SALE or AUTH")
        }
        // Create POST string to send to BluePay
        var post = ""
        post += "ACCOUNT_ID=%@"
        post += "&MODE=%@"
        post += "&TRANS_TYPE=%@"
        post += "&NAME1=%@"
        post += "&NAME2=%@"
        post += "&ADDR1=%@"
        post += "&CITY=%@"
        post += "&STATE=%@"
        post += "&ZIP=%@"
        post += "&COUNTRY=%@"
        post += "&PHONE=%@"
        post += "&EMAIL=%@"
        var postString = String(format: post, bpSetup["AccountID"]!, bpSetup["TransMode"]!, bpSetup["TransType"]!, customerInfo["FirstName"]!, customerInfo["LastName"]!, customerInfo["Street"]!, customerInfo["City"]!, customerInfo["State"]!, customerInfo["ZIP"]!, customerInfo["Country"]!, customerInfo["Phone"]!, customerInfo["Email"]!)
        post = ""
        if (customerInfo["pubKey"] != nil) {
            let tps: String = BluePay.calcTPS(bpSetup["SecretKey"]!, accID: bpSetup["AccountID"]!, transType: bpSetup["TransType"]!, transAmount: "", fullName: customerInfo["FirstName"]!, paymentAcct: "")
            post += "&TAMPER_PROOF_SEAL=%@"
            post += "&APPLE_EPK=%@"
            post += "&APPLE_DATA=%@"
            post += "&APPLE_SIG=%@"
            postString += String(format: post, tps, customerInfo["pubKey"]!.stringByAddingPercentEncodingForRFC3986(), customerInfo["data"]!.stringByAddingPercentEncodingForRFC3986(), customerInfo["signature"]!.stringByAddingPercentEncodingForRFC3986())
        }
        else if (customerInfo["EncryptedTrack1"] != nil) {
            let tps: String = BluePay.calcTPS(bpSetup["SecretKey"]!, accID: bpSetup["AccountID"]!, transType: bpSetup["TransType"]!, transAmount: customerInfo["Amount"]!, fullName: customerInfo["FirstName"]!, paymentAcct: "")
            post += "&TAMPER_PROOF_SEAL=%@"
            post += "&AMOUNT=%@"
            post += "&TRACK1_ENC=%@"
            post += "&TRACK1_EDL=%@"
            post += "&KSN=%@"
            postString += String(format: post, tps, customerInfo["Amount"]!, customerInfo["EncryptedTrack1"]!, customerInfo["Track1Length"]!, customerInfo["KSN"]!)
        }
        else {
            let tps: String = BluePay.calcTPS(bpSetup["SecretKey"]!, accID: bpSetup["AccountID"]!, transType: bpSetup["TransType"]!, transAmount: customerInfo["Amount"]!, fullName: customerInfo["FirstName"]!, paymentAcct: customerInfo["CardNumber"]!)
            post += "&TAMPER_PROOF_SEAL=%@"
            post += "&AMOUNT=%@"
            post += "&PAYMENT_ACCOUNT=%@"
            post += "&CARD_EXPIRE=%@"
            postString += String(format: post, tps, customerInfo["Amount"]!, customerInfo["CardNumber"]!, customerInfo["CardExpirationDate"]!)
        }
            // POST to the bp20post API of the BluePay gateway
        let postURL: String = "https://secure.bluepay.com/interfaces/bp20post"
        let postRequest: NSMutableURLRequest = NSMutableURLRequest(URL: NSURL(string: postURL)!)
        postRequest.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        postRequest.setValue("BluePay iOS SDK", forHTTPHeaderField: "User-Agent")
        postRequest.cachePolicy = NSURLRequestCachePolicy.ReloadIgnoringCacheData
        // Create the POST request and specify its body data
        postRequest.HTTPMethod = "POST"
        
        postRequest.HTTPBody = postString.dataUsingEncoding(NSUTF8StringEncoding)
        let task = NSURLSession.sharedSession().dataTaskWithRequest(postRequest) { data, response, error in
            guard error == nil && data != nil else {                                                          // check for fundamental networking error
                postCompleted(succeeded: false, msg: "error=\(error)")
                return
            }
            if let httpStatus = response as? NSHTTPURLResponse where httpStatus.statusCode != 200 {           // check for http errors
                //print("statusCode should be 200, but is \(httpStatus.statusCode)")
                //print("response = \(response)")
            }
            
            let responseString = NSString(data: data!, encoding: NSUTF8StringEncoding)
            postCompleted(succeeded: true, msg: responseString as! String)
        }
        task.resume()
        return
        // Return the body of the HTTPS POST response from BluePay
        //return responseBody
    }
}